src/panpkg/api.nim

```diff
@@ -180,6 +176,34 @@ proc loadImage*(path: string): Image =
     stride = result.width.cint * 4,
   )
 
+proc rgbaToArgb(img: var seq[byte], width, height, channels: int, premultiply: bool) =
+  {.push checks: off.}
+
+  for y in 0..<height:
+    for x in 0..<width:
+      let
+        r = channels * (x + y * width)
+        g = r + 1
+        b = g + 1
+        a = b + 1
+      if premultiply:
+        let
```

Comment:
I wonder if the compiler will hoist this `if` expression out of the loop as an optimization, have you checked this? Checking the condition each iteration would be disastrous for performance.

To ensure this optimization is always there, we can create a `template` out of this loop with a `static bool` parameter. like so:
```nim
proc rgbaToArgb(img: var seq[byte], width, height, channels: int, premultiply: bool) =
  {.push checks: off.}

  template conversionLoop(premultiply: static bool) {.dirty.} =
    for y in 0..<height:
      for x in 0..<width:
        discard "TODO"  # and then check `premultiply` here using `when` instead of `if`

  if premultiply:
    conversionLoop(premultiply = true)
  else:
    conversionLoop(premultiply = false)
```

Reply:
Ahh great. I actually looked into trying to do that with a template but my newness with Nim and the docs let me down and I couldn't work out how to do it. Definitely a good improvement. 

I didn't measure the performance I figured it wouldn't be too bad since it was only going to be incurred once when the image is loaded at the start of the script but this is certainly nicer.

Reply:
I tried making this change ([see here](https://github.com/liquidev/pan/commit/e732c52f012e5acb07254ace625fffd576543603)) but I'm getting this error that I don't understand. Is it saying that the template doesn't have access to the arguments of `rgbaToArgb` and need to be passed to it?

```
/home/wmoore/Source/github.com/wezm/pan/src/panpkg/api.nim(202, 19) template/generic instantiation of `conversionLoop` from here
/home/wmoore/Source/github.com/wezm/pan/src/panpkg/api.nim(186, 29) Error: type mismatch: got <seq[byte], proc [*missing parameters*](v`gensym1647: Vec[r.N, r.T]): T{.inline.} | proc [*missing parameters*](v`gensym1647: var Vec[r.N, r.T]): var T{.inline.}>
but expected one of:
proc `[]`(n: NimNode; i: BackwardsIndex): NimNode
  first type mismatch at position: 0
proc `[]`(n: NimNode; i: int): NimNode
  first type mismatch at position: 0
proc `[]`(s: string; i: BackwardsIndex): char
  first type mismatch at position: 0
proc `[]`(s: var string; i: BackwardsIndex): var char
  first type mismatch at position: 0
proc `[]`[I: Ordinal; T](a: T; i: I): T
  first type mismatch at position: 0
proc `[]`[Idx, T; U, V: Ordinal](a: array[Idx, T]; x: HSlice[U, V]): seq[T]
  first type mismatch at position: 0
proc `[]`[Idx, T](a: array[Idx, T]; i: BackwardsIndex): T
  first type mismatch at position: 0
proc `[]`[Idx, T](a: var array[Idx, T]; i: BackwardsIndex): var T
  first type mismatch at position: 0
proc `[]`[N, T](v: Vec[N, T]; ix: int): T
  first type mismatch at position: 0
proc `[]`[N, T](v: var Vec[N, T]; ix: int): var T
  first type mismatch at position: 0
proc `[]`[T, U: Ordinal](n: NimNode; x: HSlice[T, U]): seq[NimNode]
  first type mismatch at position: 0
proc `[]`[T, U: Ordinal](s: string; x: HSlice[T, U]): string
  first type mismatch at position: 0
proc `[]`[T; U, V: Ordinal](s: openArray[T]; x: HSlice[U, V]): seq[T]
  first type mismatch at position: 0
proc `[]`[T](s: openArray[T]; i: BackwardsIndex): T
  first type mismatch at position: 0
proc `[]`[T](s: var openArray[T]; i: BackwardsIndex): var T
  first type mismatch at position: 0
template `[]`(s: string; i: int): char
  first type mismatch at position: 0

expression: `[]`(img, r)
       Tip: 21 messages have been suppressed, use --verbose to show them.
     Error: Build failed for package: pan
        ... Execution failed with exit code 1
        ... Command: /usr/bin/nim c --colors:on --noNimblePath -d:NimblePkgVersion=0.2.0 --path:/home/wmoore/.nimble/pkgs/cairo-1.1.1 --path:'/home/wmoore/.nimble/pkgs/rapid-#21c847e' --path:/home/wmoore/.nimble/pkgs/aglet-0.4.4 --path:/home/wmoore/.nimble/pkgs/glm-1.1.1 --path:/home/wmoore/.nimble/pkgs/glfw-4.0.0 --path:/home/wmoore/.nimble/pkgs/stb_image-2.5 --path:/home/wmoore/.nimble/pkgs/nimLUA-0.3.8 --path:/home/wmoore/.nimble/pkgs/stb_image-2.5 --path:'/home/wmoore/.nimble/pkgs/weave-#5034793' --path:/home/wmoore/.nimble/pkgs/synthesis-0.2.0 --hints:off -o:/home/wmoore/Source/github.com/wezm/pan/pan /home/wmoore/Source/github.com/wezm/pan/src/pan.nim
```

Reply:
Gosh, that's a terrible error message... It _is_ possible some template hygiene rules changed since I last used Nim, I'd just try passing the parameters explicitly to the template at this point. Unfortunately a lot of Nim metaprogramming is just trial and error to see what works :/

Reply:
I worked it out. In doing the changes I accidentally dropped these lines:

```nim
        let
          r = channels * (x + y * width)
          g = r + 1
          b = g + 1
          a = b + 1
```

so the subsequent seq indexes were trying to use variables that didn't exist... the compiler did not make that very obvious though 😅